packages/theme-language-server-common/src/css/CSSLanguageService.ts

```diff
@@ -0,0 +1,129 @@
+
+export class CSSLanguageService {
+  private services: Record<Mode, LanguageService | null>;
+
+  constructor(
+    private documentManager: DocumentManager,
+    private getModeForURI: (uri: string) => Promise<Mode>,
+  ) {
+    this.services = { [Modes[0]]: null, [Modes[1]]: null };
+  }
+
+  async setup(clientCapabilities: LSPClientCapabilities) {
+    await Promise.all(
+      Modes.map(async (mode) => {
+        const service = getCSSLanguageService({
+          clientCapabilities,
+        });
+
+        this.services[mode] = service;
+      }),
+    );
+  }
+
+  async completions(params: CompletionParams): Promise<null | CompletionList | CompletionItem[]> {
+    const mode = await this.getModeForURI(params.textDocument.uri);
+    const service = this.services[mode];
+    if (!service) return null;
+    const documents = this.getDocuments(params, service);
+    if (!documents) return null;
+    const [stylesheetTextDocument, stylesheetDocument] = documents;
+    return service.doComplete(stylesheetTextDocument, params.position, stylesheetDocument);
+  }
+
+  async diagnostics(params: DocumentDiagnosticParams): Promise<Diagnostic[]> {
+    const mode = await this.getModeForURI(params.textDocument.uri);
+    const service = this.services[mode];
+    if (!service) return [];
+    const documents = this.getDocuments(params, service);
+    console.log(documents);
+    if (!documents) return [];
+    const [stylesheetTextDocument, stylesheetDocument] = documents;
+    return service.doValidation(stylesheetTextDocument, stylesheetDocument);
+  }
+
+  async hover(params: HoverParams): Promise<Hover | null> {
+    const mode = await this.getModeForURI(params.textDocument.uri);
+    const service = this.services[mode];
+    if (!service) return null;
+    const documents = this.getDocuments(params, service);
+    if (!documents) return null;
+    const [stylesheetTextDocument, stylesheetDocument] = documents;
+    return service.doHover(stylesheetTextDocument, params.position, stylesheetDocument);
+  }
+
+  private getDocuments(
+    params: HoverParams | CompletionParams | DocumentDiagnosticParams,
+    service: LanguageService,
+  ): [TextDocument, Stylesheet] | null {
+    const document = this.documentManager.get(params.textDocument.uri);
+    if (!document) return null;
+
+    switch (document.type) {
+      case SourceCodeType.JSON: {
+        return null;
+      }
+      case SourceCodeType.LiquidHtml: {
+        if (document.ast instanceof Error) return null;
+        const textDocument = document.textDocument;
+        let offset = 0;
+        let isDiagnostics = false;
+        if ('position' in params && params.position.line !== 0 && params.position.character !== 0) {
+          offset = textDocument.offsetAt(params.position);
+        } else {
+          const stylesheetIndex = document.source.indexOf('{% stylesheet %}');
+          console.log(stylesheetIndex);
+          offset = stylesheetIndex;
+          isDiagnostics = true;
```

Comment:
Why not use the visitor option?
```ts
const stylesheet = ancestors.find(
  (node): node is LiquidRawTag =>
    node.type === NodeTypes.LiquidRawTag && node.name === 'stylesheet',
);
```

I _believe_ the stylesheet tag is also of NodeType.LiquidRawTag so that the prettier plugin can switch to the css embed mode for those.

Reply:
oh derp two lines down. I should read lol